napari/_qt/_tests/test_qt_viewer.py

```diff
@@ -657,3 +659,14 @@ def test_insert_layer_ordering(make_napari_viewer):
     pl2_vispy = viewer.window._qt_viewer.layer_to_visual[pl2].node
     assert pl1_vispy.order == 1
     assert pl2_vispy.order == 0
+
+
+def test_create_non_empty_viewer_model(qtbot):
+    viewer_model = ViewerModel()
+    viewer_model.add_points([(1, 2), (2, 3)])
+
+    viewer = QtViewer(viewer=viewer_model)
+
+    viewer.close()
+    viewer.deleteLater()
+    qtbot.wait(50)
```

Comment:
I think we do this elsewhere and I suspect this is related to the lifetime of C++/Python objects and some of the errors we see associated with that.

To help clarify, what do we hope will happen in the 50ms we wait here? I.e. what are we waiting for? I've seen [the comment in `_testsupport.py`](https://github.com/napari/napari/blob/bdbb331884a6b22c8178dfb796d6620fe76a996d/napari/utils/_testsupport.py#L173), but that doesn't explain why we do this. And why we don't use `qtbot.addWidget(viewer)` instead/in addition.

Reply:
Because if You add the object to `qtbot`, there is no guarantee that the object is cleaned before starting the next test. So `QtViewer` is detected on begin of the next test that causes its failiture. 

Waiting is required to give Qt time to clean the object. Qt doesn't process deletion (and connected with its signal processing) immediately but waits on the moment when there are free resources. This allows to have better program performance (no lag in render interface), but conflicts with napari detection of leaked widgets.

Reply:
Thanks for the explanation. And sorry for discussing this in this PR - we can take this conversion somewhere else if you'd like.

Just to confirm, waiting 50ms doesn't guarantee that the object is cleaned properly either, but makes it more likely, right? By cleaned, do you mean destroyed (i.e. the C++ QObject destructor has finished)? If so, could we wait for [the `QObject::destroyed` signal](https://doc.qt.io/qt-5/qobject.html#destroyed)?

Also, why do we care if there are any existing `QtViewer` objects around when we start the new test? Is it related to the lifetime of the `QApplication`?

Reply:
> Just to confirm, waiting 50ms doesn't guarantee that the object is cleaned properly either, but makes it more likely, right? By cleaned, do you mean destroyed (i.e. the C++ QObject destructor has finished)? If so, could we wait for [the `QObject::destroyed` signal](https://doc.qt.io/qt-5/qobject.html#destroyed)?

Yes. If I good remember waiting on destroyed does not work. I do not remember why. We could check again.
 
> Also, why do we care if there are any existing `QtViewer` objects around when we start the new test? Is it related to the lifetime of the `QApplication`?

I do not know. This is a question to @tlambert03

Reply:
Yeah I vaguely recall this one as just being hard to figure out.  Would love to drop the use of wait if we can figure it out... was probably just a "surrender" here.

I don't think we specifically care about having one around when when another one starts actually.  The original "leaked widget" test there was more about _cleanup_ to prevent sneaky accumulation of memory leaks: by making sure that when we close the last viewer window, there are no more stray Qt objects laying around that haven't been properly patented.

It's rather annoying of course :) but that's the goal

Reply:
Thanks for the info. From the Qt C++ side, I don't think we should have to worry too much. Either closing will trigger deletion if `Qt.WidgetAttribute. WA_DeleteOnClose` is set (which it often is for napari widgets), or Qt will handle deletion through the parent/child relationships assuming they have been setup correctly.

For this test specifically, I'm skeptical that `deleteLater` is doing anything because we set [`Qt.WidgetAttribute.WA_DeleteOnClose` on the `QtViewer` instance](https://github.com/napari/napari/blob/87a263328fd71c8d3fc6252fe8ee2d09056ae776/napari/_qt/qt_viewer.py#L191). If I'm reading [the Qt docs correctly](https://doc.qt.io/qt-5/qt.html#WidgetAttribute-enum), then I think `QObject::delete` will be called after [the close event has been accepted in `QtViewer.closeEvent`](https://github.com/napari/napari/blob/87a263328fd71c8d3fc6252fe8ee2d09056ae776/napari/_qt/qt_viewer.py#L1139).

I'm less sure about the Python objects, as that's what `QtViewer` has weak references to. Specifically, I'm not sure when pyqt/pysidep will cause their reference counts to drop to 0 (i.e. are we the only ones holding references to them?).

But we should probably discuss this elsewhere (maybe Zulip?), as we already use this pattern elsewhere in the tests, so this PR doesn't do anything worse than `main` and also fixes a bug.

Reply:
> or Qt will handle deletion through the parent/child relationships assuming they have been setup correctly.

this is exactly the assumption that is being tested here.  With a lot of developers who don't have Qt experience, we saw a lot of parentless widgets leaking through, so after fixing that, this got set up to prevent a regression